cirq/experiments/fidelity_estimation.py

```diff
@@ -372,3 +394,225 @@ def least_squares_xeb_fidelity_from_probabilities(
     return least_squares_xeb_fidelity_from_expectations(
         measured_expectations, exact_expectations, uniform_expectations
     )
+
+
+@dataclass(frozen=True)
+class _Sample2qXEBTask:
+    """Helper contained for grouping a circuit to be sampled."""
+
+    cycle_depth: int
+    circuit_i: int
+    prepared_circuit: 'cirq.Circuit'
+
+
+class _SampleInBatches:
+    def __init__(self, sampler: 'cirq.Sampler', repetitions: int):
+        """This closure will execute a list of `tasks` with one call to
+        `run_batch` on the provided sampler for a given number of repetitions."""
+        self.sampler = sampler
+        self.repetitions = repetitions
+
+    def __call__(self, tasks: List[_Sample2qXEBTask]):
+        prepared_circuits = [task.prepared_circuit for task in tasks]
+        results = self.sampler.run_batch(prepared_circuits, repetitions=self.repetitions)
+        assert len(results) == len(tasks)
+        records = []
+        for task, n_result in zip(tasks, results):
+            assert len(n_result) == 1, len(n_result)
+            result = n_result[0]  # sweep nesting
+            sampled_inds = result.data.values[:, 0]
+            sampled_probs = np.bincount(sampled_inds, minlength=2 ** 2) / len(sampled_inds)
+
+            records += [
+                {
+                    'circuit_i': task.circuit_i,
+                    'cycle_depth': task.cycle_depth,
+                    'sampled_probs': sampled_probs,
+                }
+            ]
+        return records
+
+
+def sample_2q_xeb_circuits(
+    sampler: 'cirq.Sampler',
+    circuits: Sequence['cirq.Circuit'],
+    cycle_depths: Sequence[int],
+    repetitions: int = 10_000,
+    batch_size: int = 9,
+):
+    """Sample two-qubit XEB circuits given a sampler.
+
+    Args:
+        sampler: A Cirq sampler for executing circuits.
+        circuits: A library of two-qubit circuits generated from
+            `random_rotations_between_two_qubit_circuit` of sufficient length
+            for `cycle_depths`.
+        cycle_depths: A sequence of cylce depths at which we will truncate each
+            of the `circuits` to execute.
+        repetitions: Each (circuit, cycle_depth) will be sampled for this many
+            repetitions.
+        batch_size: We call `run_batch` on the sampler, which can speed up
+            execution in certain environments. The number of
+            (circuit, cycle_depth) tasks to be run in each batch
+            is given by this number.
+
+    Returns:
+        A pandas dataframe with index given by ['circuit_i', 'cycle_depth'] and
+        column "sampled_probs".
+    """
+    import tqdm
+
+    all_all_qubits_s: Set['cirq.Qid'] = set()
```

Comment:
half of these variable names are because mypy doesn't let you re-declare a variable with a different type which is very annoying. I'll try to make them more readable

Reply:
Went through and gave full names. Also added a docstring to `_Sample2qXEBTask` to describe what is meant by `prepared_circuit`.

Reply:
Much better